saturn_client/core.py

```diff
@@ -0,0 +1,161 @@
+"""Python library for interacting with the Saturn Cloud API.
+
+NOTE: This is an experimental library and will likely change in
+the future
+"""
+
+import json
+import logging
+
+from typing import Any, Dict, Optional
+from urllib.parse import urljoin
+
+import requests
+from .settings import Settings
+
+
+log = logging.getLogger("saturn-client")
+if log.level == logging.NOTSET:
+    log.setLevel(logging.INFO)
+
+
+class SaturnConnection:
+    """
+    Create a ``SaturnConnection`` to interact with the API.
+
+    :param url: URL for the SaturnCloud instance.
+    :param api_token: API token for authenticating the request to Saturn API.
+        Get from `/api/user/token`
+    """
+
+    _options = None
+    _saturn_api_version = None
+
+    def __init__(
+        self,
+        url: Optional[str] = None,
+        api_token: Optional[str] = None,
+    ):
+        """
+        Create a ``SaturnConnection`` to interact with the API.
+
+        :param url: URL for the SaturnCloud instance.
+        :param api_token: API token for authenticating the request to Saturn API.
+            Get from `/api/user/token`
+        """
+        self.settings = Settings(url, api_token)
+
+        # test connection to raise errors early
+        self._saturn_api_version = self.saturn_api_version
+
+    @property
+    def url(self):
+        """URL of Saturn instance"""
+        return self.settings.url
+
+    @property
+    def saturn_api_version(self):
+        """Version of Saturn API"""
+        if self._saturn_api_version is None:
+            url = urljoin(self.url, "api/status")
+            response = requests.get(url, headers=self.settings.headers)
+            if not response.ok:
+                raise ValueError(response.reason)
+            self._saturn_api_version = response.json()["version"]
+        return self._saturn_api_version
+
+    @property
+    def options(self):
+        """Options for various settings"""
+        if self._options is None:
+            url = urljoin(self.url, "api/info/servers")
+            response = requests.get(url, headers=self.settings.headers)
+            if not response.ok:
+                raise ValueError(response.reason)
+            self._options = response.json()
+        return self._options
+
+    @property
+    def describe_sizes(self) -> Dict[str, str]:
+        """Instance size options"""
+        return self.options["sizes"]
+
+    def create_project(
+        self,
+        name: Optional[str] = None,
```

Comment:
`name` should not be optional. If this client is going to reference resources in Saturn by `name` (it seems that way), I don't think it's unreasonable to say that you have to provide one.

Reply:
Yeah this was just a mistake. Name is required on the backend.